MAINT: Use the same exception for all bad axis requests

eric-wieser opened the pull request:
Lots of functions accept an axis argument that can be negative, implying wraparound. The same few lines of code are repeated all over the place to implement this.

When the axis is just plain invalid, an error is raised. However:
* Some things raise `IndexError`, others raise `ValueError`. This PR somehwat arbitrarily assumes that `IndexError` is the right type of exception to raise.
* Some messages are more useful than others
* Some functions don't check for `axis < -ndim`

It seems it would be better to put all this in one place.

This was previously filed as an issue at  #8583.   Function name and location inspired by `check_and_adjust_index`. `check_and_adjust_index(...) < 0` is used to match `check_and_adjust_axis(...) < 0`.

This was previously discussed [on the mailing list](https://mail.scipy.org/pipermail/numpy-discussion/2016-September/075951.html) by @charris and @seberg

eric-wieser committed:
MAINT: Use the same error message for most C code handling negative axes

eric-wieser committed:
ENH: Expose np.core.multiarray.normalize_axis_index(axis, ndim)

mhvk commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -263,13 +263,7 @@ PyArray_ConvertMultiAxis(PyObject *axis_in, int ndim, npy_bool *out_axis_flags)
-                PyErr_Format(PyExc_ValueError,
-                        "'axis' entry %d is out of bounds [-%d, %d)",
-                        axis_orig, ndim, ndim);
+            if (check_and_adjust_axis(&axis, ndim) < 0) {
```
Here and below: check for status using `!= 0`? Or just omit the comparison altogether (as above in `error_converting(axis)`

eric-wieser replied:
Again, written this way for consistency with `check_and_adjust_index`

mhvk commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -306,11 +297,8 @@ PyArray_ConvertMultiAxis(PyObject *axis_in, int ndim, npy_bool *out_axis_flags)
-                    axis_orig, ndim, ndim);
-            return NPY_FAIL;
+        if (check_and_adjust_axis(&axis, ndim) < 0) {
+            return NULL;
```
Shouldn't one still return `NPY_FAIL` here?

eric-wieser replied:
Good catch

mhvk commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -1104,13 +1104,10 @@ PyArray_Sort(PyArrayObject *op, int axis, NPY_SORTKIND which)
-        PyErr_Format(PyExc_ValueError, "axis(=%d) out of bounds", axis_orig);
-        return -1;
+    if (check_and_adjust_axis(&axis, n) < 0) {
+        return NULL;
```
Same here: I think the return argument shouldn't change, i.e., return `-1` here?

mhvk commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -1216,13 +1213,10 @@ PyArray_Partition(PyArrayObject *op, PyArrayObject * ktharray, int axis,
-        PyErr_Format(PyExc_ValueError, "axis(=%d) out of bounds", axis_orig);
-        return -1;
+    if (check_and_adjust_axis(&axis, n) < 0) {
+        return NULL;
```
idem

mhvk commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -1455,13 +1449,8 @@ PyArray_LexSort(PyObject *sort_keys, int axis)
-                "axis(=%d) out of bounds", axis);
-        goto fail;
+    if (check_and_adjust_axis(&axis, nd) < 0) {
+        return NULL;
```
Definitely should be `goto fail` here, since several `DECREF`s need to be done.

mhvk commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4109,6 +4103,24 @@ array_may_share_memory(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *
     return array_shares_memory_impl(args, kwds, NPY_MAY_SHARE_BOUNDS, 0);
 }
 
+static PyObject *
```
I presume this is for follow up in python code? If so, I think it should be part of a follow-up PR, not be done here (if only to give a chance to discuss whether it is best to go through `C` for something like this).

eric-wieser replied:
I was going to dump that in the same PR, and just hadn't finished yet

eric-wieser replied:
I've added the python stuff to this PR now...

mhvk commented:
This makes sense to me, but note that I think you'll have to ensure the code keeps behaving the same in case of failure. As you will have seen, I suggest postponing the method to a possible python follow-up PR.

There are also a few more places where the code could be used:
1. https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/item_selection.c#L1832
2. https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/shape.c#L708
3. https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/ufunc_object.c#L3983

mhvk commented:
Actually, looking more closely at anything that includes `axis.*<.*0`, I find there is already a method which would seem to do what you want... `check_and_adjust_index` -- https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/common.h#L110

mhvk commented:
p.s. Well, the function is very similar but doesn't exactly do the same thing, since that one checks an index rather than an axis, but with the same logic. I'm not sure I understand the threading stuff...

mhvk commented:
p.p.s. And I really should learn to read commit messages -- you of course refer to `check_and_adjust_index` already...

eric-wieser commented on numpy/core/src/multiarray/common.h:
```diff
@@ -134,6 +134,28 @@ check_and_adjust_index(npy_intp *index, npy_intp max_item, int axis,
+ * 0 <= *axis < ndim, and returns 0.
+ */
+static NPY_INLINE int
+check_and_adjust_axis(int *axis, int ndim)
```
Naming it this because `check_and_adjust_index` does exactly the same type of transformation.

eric-wieser committed:
MAINT: Use normalize_axis_index in all python axis checking

As a result, some exceptions change from ValueError to IndexError

This also changes the exception types raised in places where
normalize_axis_index is not quite appropriate

eric-wieser commented:
Wow, that was a clumsy bunch of mistakes. Thanks for catching me out!

> There are also a few more places where the code could be used:

Great, thanks for spotting those - I'll add a fixup commit for those shortly

> you'll have to ensure the code keeps behaving the same in case of failure. 

Meaning what? Produce the same error message?

mhvk commented:
>> you'll have to ensure the code keeps behaving the same in case of failure.

> Meaning what? Produce the same error message?

No, just that what the code returns (error code or otherwise) remains the same (see in-line comments)

eric-wieser commented:
Ok, everything is fixed except one thing.

[These lines](https://github.com/numpy/numpy/blob/7927b3cb4b57e79bf6b1b8a723583295fc0398eb/numpy/ma/core.py#L3924-L3928) appear multiple times in `numpy/ma/core.py`, such as `MaskedArray.__eq__:
```python
try:
    mask = mask.view((bool_, len(self.dtype))).all(axis)
except ValueError:
    mask = np.all([[f[n].all() for n in mask.dtype.names]
      for f in mask], axis=axis)
```

What is that ValueError trying to catch? An error in `.view`, or an error in `.all`?

mhvk commented:
I like how much the code cleans up using this! A few things I'm uncertain about:
1. For the python side, a pure python implementation would have the advantage that it is easier to see what it does (this is partially why I thought it might be good to do it in a separate PR);
2. For the python side, I'd prefer a name that makes more obvious what it does, e.g., `normalize_axis`);
3. Am not quite sure what the appropriate error is; people may count on getting `ValueError`, though I would agree that `IndexError` is more logical.

eric-wieser commented:
@mhvk 

1. I don't like the idea of implementing the same code in both C and python, leaving the only alternative as having the C code call python - but that would be a performance hit
2. I've deliberately not exposed it publically at `np.resolve_axis` because I was not sure of the name - I'd welcome some more input - I don't have a good argument for one or the other
3. I'd be in favor of making a special `AxisError` type, because that would make the tests more explicit. Of course, the question then becomes _"should this subclass `ValueError`, `IndexError`, or just `Exception`"_...

charris commented:
Grrr, github is now notifying me of commits without referencing a PR. It's an "improvement" too far.

eric-wieser commented:
@charris : that's my fault for doing some url mangling in order to comment on a commit and not a diff line

mhvk commented:
Hmm, possibly one needs `class AxisError(IndexError, ValueError)`...

mhvk commented:
More seriously: which `MaskedArray` test is actually broken by the change from `ValueError` to `IndexError` -- I tried looking at the travis results, but only noticed 2 failures, which got me confused (maybe I just looked badly).

eric-wieser commented:
@mhvk: Look in the test results _before_ the commit that claims to fix them ;)

eric-wieser commented on numpy/core/numeric.py:
```diff
@@ -1527,15 +1527,12 @@ def rollaxis(a, axis, start=0):
-        axis += n
+    axis = resolve_axis(axis, n)
     if start < 0:
         start += n
```
This has always struck me as a bug - IMO, `rollaxis(np.zeros(3,4,5), 0, -1).shape == (4, 5, 3)`, ie `-1` should mean "move this axis to the end. The fact that this is the only place that can't use `resolve_axis` sort of backs that up.

mhvk replied:
Yes, I have been caught by that too, realising only slowly that I had to put `array.ndim` if I wanted something moved to the last axis. I think this is partially why we now have `moveaxis` -- in any case, I think it is fine to leave this one untouched!

mhvk commented:
I did...  E.g., https://travis-ci.org/numpy/numpy/jobs/199817411 (2 errors only). But now I look at the appveyor results, I see that there are also 3 errors in `test_recfunctions`. Most bizarre. Maybe travis deletes jobs sooner? Or the github link is incorrect?

eric-wieser commented:
[This appveyor page](https://ci.appveyor.com/project/charris/numpy/build/1.0.4064) shows the bug

mhvk commented:
@eric-wieser - with #8590, you should not need to fix `ma/core.py` any more.

eric-wieser commented:
> you should not need to fix ma/core.py any more.

I still need to swap `ValueError` for `IndexError` for the tests to pass. But yes, the change will be overwritten when you merge that PR

shoyer commented:
I don't think `IndexError` is quite right here, because that exception is quite specifically for subscripting a sequence. I would lean towards `class AxisError(ValueError)`, or maybe `class AxisError(IndexError, ValueError)` if we're concerned about backwards compatibility.

eric-wieser commented:
> because that exception is quite specifically for subscripting a sequence

Are we not essentially subscripting `arr.shape[axis]` here though?

What would be an appropriate name for `AxisError` in the C code, and which file would it be best defined in?

mhvk commented:
I think I'd also pick `IndexError` over `ValueError` as we're indexing the array's dimensions. But I worry about breaking code in either case... That said, the fact that so little broke in numpy by doing this is a good sign (the `MaskedArray` one I'd chalk up to pure coincidence).

shoyer commented:
Sure, in many cases `axis` is used like `arr.shape[axis]` but I don't think that's always true and in any case it clearly isn't obvious to everyone -- that's how we ended up with many of these being the more generic `ValueError` in the first place.

Anyways, I think we do need to inherit from both for backwards compatibility.

Since we need to call the exception from C, it looks like we need to define it in C, too:
https://docs.python.org/2/c-api/exceptions.html#c.PyErr_NewException
http://pythonextensionpatterns.readthedocs.io/en/latest/exceptions.html

I'm not sure about doing multiple-inheritance in C, though, it looks tricky:
http://stackoverflow.com/questions/40593529/multiple-inheritance-in-python-c-api

eric-wieser committed:
ENH: For compatibility, use an exception type that subclasses both original types

eric-wieser commented:
@shoyer: Or I can take the easy route and just put it in `np.core._internal` ;)

eric-wieser committed:
MAINT: Be specific about where AxisError is raised

These were tested by temporarily removing the base classes from AxisError

eric-wieser commented:
@shoyer: Ok, all set. Needs documentation yet on the exposed function and error  but I'd appreciate some comments on the general approach.

homu commented:
:umbrella: The latest upstream changes (presumably #8441) made this pull request unmergeable. Please resolve the merge conflicts.

mhvk commented on numpy/core/src/multiarray/common.h:
```diff
@@ -134,6 +134,42 @@ check_and_adjust_index(npy_intp *index, npy_intp max_item, int axis,
+    if (NPY_UNLIKELY((*axis < -ndim) || (*axis >= ndim))) {
+        /*
+         * Load the exception type, if we don't already have it. Unfortunately
+         * we don't have access to npy_cache_import here
```
More for my education than anything else, is there a reason not just to `#include <npy_import.h>` at the top?

eric-wieser replied:
This is a public header, but I think `npy_import` is private. So when another module includes this header, it can no longer find `npy_import`.

That could be fixed by moving this to `common.c`, but I don't know if the inlining would then work

mhvk replied:
Since every problem supposedly is solvable with another layer of redirection, could one call an error-producing function in `common.c` here, where that function raises the error? (But maybe this is very much not worth it...)

eric-wieser replied:
Arguably, the inlining is not important here, as this isn't nearly as critical a path as `check_and_adjust_index`

mhvk replied:
OK, how about then moving it to `common.c`? Somewhat nicer anyway to have c code reside in `*.c` files...

eric-wieser replied:
Main argument for leaving it here is it gives clear contrast with `check_and_adjust_index`

mhvk replied:
True; fine to keep as is too even if it duplicates a bit of code.

charris replied:
@eric-wieser did you try to include `npy_import.h`? It should work. Private here means private to numpy. The files therein should be available to everything in `numpy/core/src`.

eric-wieser replied:
I think the problem is that something outside numpy/core/src tries to include "common.h", which then doesn't have access to npy_include. I remember it not building, but I forget the exact error.

charris replied:
And just to be clear, the "public" headers are in `numpy/core/include/numpy`.

mhvk commented:
Defining `AxisError` in `_internal.py` seems a good solution.

As for the name of the python function, I think I'd prefer `normalize_axis` over `resolve_axis`.

Personally, I think I'd still prefer to define it in python on the python side, but I don't feel strongly. In any case, it will need to have a docstring.

eric-wieser commented:
Do we perhaps want `normalize_axis_index` just to be clear that we're not normalizing along an axis?

mhvk commented:
`normalize_axis_index` would also be fine with me. @shoyer?

shoyer commented:
I like `normalize_axis_index`. Just to be clear, this is for naming a NumPy internal function, right? We aren't exposing this as external API?

seberg commented:
+1 for not exposing it I think, so adding an underscore is probably good.

eric-wieser commented:
Right now, it is only accessible via `np.core.multiarray.normalize_axis_index`.

@seberg, would you prefer it to be semi-public at `np._normalize_axis_index`?

eric-wieser commented:
Ok, renamed to `np.core.multiarray.normalize_axis_index`, and a docstring added.

seberg commented:
I did not follow this much, but I have a believe that someone will find any weird function and think its fine to use it. So, unless we are fine with people possibly starting to use it, adding an underscore to underscore ;) that its considered private just seems like a save bet.

eric-wieser commented:
@seberg: I'd argue that this is something we actually want people to use - for anyone writing their own function taking an `axis` argument that can't just forward it, this is a simple way to make it behave as people expect.

seberg commented:
Ok, sounds fine to me, it was more of an "if we are not sure its stable API" thing, to be honest.

mhvk commented:
I think I agree with @eric-wieser - it is good if this gets used (but would not quite yet put it at the top level)

mhvk commented:
Thanks for adding the docstring; I think all is set now, so will merge.

eric-wieser commented:
Should this have had a release note, or will that be handled during the release stage?

mhvk commented:
Hmm, yes, probably big enough to have a release note... Now best to do separately?

charris commented:
Yes, should have a release note as it is a behavior change. The type warning for the `check_and_adjust_axis` call also needs a fix that would be good to get in before it is forgotten.

charris commented:
@mhvk It's in, so will need a separate PR for the release note.

pv commented:
This appears to cause a crash in scipy, gh-8671

pv commented on numpy/core/src/multiarray/shape.c:
```diff
@@ -705,12 +705,7 @@ PyArray_Transpose(PyArrayObject *ap, PyArray_Dims *permute)
-            if (axis < 0 || axis >= PyArray_NDIM(ap)) {
-                PyErr_SetString(PyExc_ValueError,
-                                "invalid axis for this array");
+            if (check_and_adjust_axis(&axis, PyArray_NDIM(ap)) < 0) {
```
`axis` is `npy_intp`, not `int`.

pv replied:
Maybe useful to double check this everywhere.

eric-wieser replied:
Which do you think I should accept in `check_and_adjust_axis`? Sometimes axis can be `int`, sometimes it is `intp`

charris commented:
@eric-wieser Time to get going on that fix.

eric-wieser commented:
@charris: Will do that tomorrow
